File: tests/generate_full_video.js
#!/usr/bin/env node
/**
 * Plan helpers for the 4-fighter long-video smoke.
 *
 *   node tests/generate_full_video.js init <state.json>
 *   node tests/generate_full_video.js next <state.json> <match.json>
 *   node tests/generate_full_video.js apply <state.json> <winnerName>
 */
'use strict';

const fs = require('fs');
const path = require('path');

const FUN = path.join(__dirname, '..');
const STARTED = Date.now();

function log(kind, msg) {
    const elapsed = ((Date.now() - STARTED) / 1000).toFixed(1);
    console.error(`${new Date().toISOString()}  +${elapsed}s  ${kind}  ${msg}`);
}

global.window = global;
require(path.join(FUN, 'premade-powerups/registry.js'));
require(path.join(FUN, 'premade-powerups/power.js'));
require(path.join(FUN, 'premade-powerups/speed.js'));
require(path.join(FUN, 'premade-powerups/size.js'));
require(path.join(FUN, 'premade-powerups/thorns.js'));
require(path.join(FUN, 'premade-powerups/protection.js'));
require(path.join(FUN, 'premade-powerups/apple.js'));
require(path.join(FUN, 'premade-powerups/index.js'));
require(path.join(FUN, 'workflow/bracket.js'));
require(path.join(FUN, 'workflow/match-compose.js'));
require(path.join(FUN, 'workflow/powerup-wheel.js'));

const roster = [
    { weaponId: 'hammer', name: 'Hammer', color: '#f97316' },
    { weaponId: 'dagger', name: 'Dagger', color: '#3b82f6' },
    { weaponId: 'sword', name: 'Sword', color: '#ef4444' },
    { weaponId: 'spikes', name: 'Spikes', color: '#a855f7' },
].map((row, index) => ({
    id: '_weapon',
    name: row.name,
    color: row.color,
    weaponId: row.weaponId,
    slotKey: `slot-${index}:_weapon:${row.weaponId}`,
    slotIndex: index,
    arenaMatchup: {
        id: '_weapon',
        config: {
            weaponId: row.weaponId,
            name: row.name,
            color: row.color,
        },
    },
}));

function readJson(file) {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
}

function writeJson(file, data) {
    fs.writeFileSync(file, JSON.stringify(data));
}

function buildMatchRequest(state, order) {
    const match = WorkflowBracket.currentMatch(state);
    if (!match) throw new Error('no open match');
    const bracketPre = WorkflowBracket.clone(state);
    const spinA = PowerupWheel.createSpin({ fighter: match.a, delayMs: 0 });
    const spinB = PowerupWheel.createSpin({
        fighter: match.b,
        delayMs: PowerupWheel.nextSpinDelayMs(spinA),
    });
    PowerupWheel.applyResultToFighter(match.a, spinA.resultId);
    PowerupWheel.applyResultToFighter(match.b, spinB.resultId);
    const spins = {
        a: PowerupWheel.serializeSpin(spinA),
        b: PowerupWheel.serializeSpin(spinB),
    };
    const script = TournamentCompose.buildMatchScript({
        a: match.a,
        b: match.b,
        winner: null,
        mode: 'weapon',
        spins,
    });
    return {
        matchKey: WorkflowBracket.matchComposeKey(match),
        script,
        order,
        mode: 'weapon',
        matchup: WorkflowBracket.matchArenaMatchup(match),
        aName: TournamentCompose.plainFighterName(match.a),
        bName: TournamentCompose.plainFighterName(match.b),
        winnerName: null,
        loserName: null,
        bracketPre,
        bracketPost: null,
        activeMatch: match,
        lastWinner: null,
        lastLoser: null,
        powerupSpins: spins,
        syntheticArena: false,
        spinSummary: `${match.a.name}=${spinA.resultName || 'No powerup'} | ${match.b.name}=${spinB.resultName || 'No powerup'}`,
    };
}

function cmdInit(statePath) {
    const state = WorkflowBracket.buildFromFighters(roster);
    writeJson(statePath, {
        state,
        order: 0,
        matchKeys: [],
        championName: null,
    });
    log('STEP', `init ${roster.map((f) => f.name).join(', ')}`);
}

function cmdNext(statePath, matchPath) {
    const bag = readJson(statePath);
    const state = bag.state;
    if (state.complete && state.champion) {
        bag.championName = state.champion.name;
        writeJson(statePath, bag);
        log('STEP', `bracket complete · champion ${state.champion.name}`);
        process.exit(2);
    }
    const req = buildMatchRequest(state, bag.order);
    writeJson(matchPath, req);
    log('STEP', `next ${bag.order + 1} ${req.aName} vs ${req.bName} · ${req.spinSummary}`);
}

function cmdApply(statePath, winnerName) {
    const bag = readJson(statePath);
    const decided = WorkflowBracket.applyWinnerName(bag.state, winnerName);
    bag.order += 1;
    bag.matchKeys = bag.matchKeys || [];
    const key = WorkflowBracket.matchComposeKey(decided);
    if (key) bag.matchKeys.push(key);
    if (bag.state.complete && bag.state.champion) {
        bag.championName = bag.state.champion.name;
    }
    writeJson(statePath, bag);
    log('STEP', `apply ${winnerName} → ${decided.winner?.name}`);
}

const cmd = process.argv[2] || 'init';
try {
    if (cmd === 'init') {
        cmdInit(process.argv[3] || path.join(__dirname, '_full_video_state.json'));
    } else if (cmd === 'next') {
        cmdNext(
            process.argv[3] || path.join(__dirname, '_full_video_state.json'),
            process.argv[4] || path.join(__dirname, '_full_video_match.json'),
        );
    } else if (cmd === 'apply') {
        cmdApply(
            process.argv[3] || path.join(__dirname, '_full_video_state.json'),
            process.argv.slice(4).join(' '),
        );
    } else {
        throw new Error(`unknown command ${cmd}`);
    }
} catch (err) {
    log('ERROR', err.message || String(err));
    process.exit(1);
}

File: tests/generate_full_video.py
#!/usr/bin/env python3
"""Record a 4-fighter tournament and stitch tournament-final.mp4.

Advances the bracket from the recorded arena winner (not a demo pick).

    python3 tests/generate_full_video.py
"""

from __future__ import annotations

import json
import subprocess
import sys
import time
from pathlib import Path

FUN_DIR = Path(__file__).resolve().parents[1]
PIPELINE_DIR = FUN_DIR / "pipeline"
sys.path.insert(0, str(PIPELINE_DIR))

import compose_tournament as ct  # noqa: E402
import tournament_record as tr  # noqa: E402

STARTED = time.time()
BASE_URL = "http://127.0.0.1:8764"
PLANNER = FUN_DIR / "tests" / "generate_full_video.js"


def log(kind: str, msg: str) -> None:
    elapsed = time.time() - STARTED
    print(f"{time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())}  +{elapsed:.1f}s  {kind}  {msg}", flush=True)


def node(*args: str, check: bool = True) -> subprocess.CompletedProcess:
    return subprocess.run(
        ["node", str(PLANNER), *args],
        check=check,
    )


def main() -> int:
    state_path = FUN_DIR / "tests" / "_full_video_state.json"
    match_path = FUN_DIR / "tests" / "_full_video_match.json"
    log("STEP", "init bracket")
    node("init", str(state_path))

    log("STEP", "clearing previous tournament media")
    ct.clear_tournament_media()

    match_keys = []
    while True:
        nxt = node("next", str(state_path), str(match_path), check=False)
        if nxt.returncode == 2:
            break
        if nxt.returncode != 0:
            raise RuntimeError("planner next failed")
        req = json.loads(match_path.read_text())
        key = req["matchKey"]
        match_keys.append(key)
        log("STEP", f"match {req['order'] + 1} {req['aName']} vs {req['bName']} · {req.get('spinSummary')}")
        result = tr.ensure_match_segment_media(
            match_key=key,
            script=req["script"],
            order_index=int(req["order"]),
            mode=req.get("mode") or "weapon",
            matchup=req["matchup"],
            a_name=req.get("aName"),
            b_name=req.get("bName"),
            winner_name=req.get("winnerName"),
            loser_name=req.get("loserName"),
            base_url=BASE_URL,
            force=True,
            synthetic_arena=False,
            bracket_pre=req.get("bracketPre"),
            bracket_post=req.get("bracketPost"),
            active_match=req.get("activeMatch"),
            last_winner=req.get("lastWinner"),
            last_loser=req.get("lastLoser"),
            powerup_spins=req.get("powerupSpins"),
        )
        winner = result.get("winnerName") or (result.get("arena") or {}).get("winner")
        if not winner:
            raise RuntimeError(f"arena produced no winner for {key}")
        seg = result.get("segment") or {}
        log("STEP", f"segment done → {seg.get('file')} ({seg.get('duration')}s) winner={winner}")
        node("apply", str(state_path), str(winner))

    bag = json.loads(state_path.read_text())
    champion = bag.get("championName") or (bag.get("state") or {}).get("champion", {}).get("name")
    if not champion:
        raise RuntimeError("bracket did not produce a champion")
    log("STEP", f"champion outro + stitch · {champion}")
    fighters = ((bag.get("state") or {}).get("fighters") or [])
    names = [
        f.get("name")
        for f in fighters
        if isinstance(f, dict) and f.get("name")
    ]
    intro_title = ct.build_intro_title(names, weapon_mode=True, entrant_count=len(names))
    intro = ct.ensure_intro_clip(
        names,
        powerup_spin=True,
        title=intro_title,
        fighters=fighters,
        champion_name=champion,
        base_url=BASE_URL,
        weapon_mode=False,
    )
    outro = ct.ensure_outro_clip(
        champion,
        title=intro_title,
        fighters=fighters,
        weapon_mode=False,
        base_url=BASE_URL,
    )
    stitched = ct.stitch_final(
        intro_clip=intro,
        champion_clip=outro,
        champion_name=champion,
        force=True,
        expected_count=len(match_keys),
        match_keys=match_keys,
    )
    path = Path(stitched.get("path") or ct.FINAL_PATH)
    size = path.stat().st_size if path.is_file() else 0
    log("DONE", f"final={stitched.get('final')} bytes={size} segments={stitched.get('segmentCount')} path={path}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:  # noqa: BLE001
        log("ERROR", str(exc))
        raise
